### ISSUE TITLE
test_subquery_exclude_outerref sometimes fails: internal error: column 6 not in input

### ISSUE THREAD

**Issue by timgraham** [cockroachdb/django-cockroachdb#139]
I've seen this failure on TeamCity with CockroachDB 19.2.x and 20.1.1 but haven't been able to reproduce locally.
```
======================================================================
ERROR: test_subquery_exclude_outerref (queries.tests.ExcludeTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/agent/work/323e0f636337ffe3/_django_repo/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.InternalError_: internal error: column 6 not in input
DETAIL:  stack trace:
/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:102: getColumnOrdinal()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:1300: buildLookupJoin()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:217: buildRelational()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:657: buildApplyJoin()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:309: buildRelational()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:1186: buildLimitOffset()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:208: buildRelational()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:569: buildProject()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:199: buildRelational()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:142: build()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:111: Build()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:178: makeOptimizerPlan()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:801: makeExecPlan()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:691: dispatchToExecutionEngine()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:481: execStmtInOpenState()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:96: execStmt()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1368: execCmd()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1297: run()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:476: ServeConn()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:590: func1()
/usr/local/go/src/runtime/asm_amd64.s:1357: goexit()

HINT:  You have encountered an unexpected error.

Please check the public issue tracker to check whether this problem is
already tracked. If you cannot find it there, please report the error
with details by creating a new issue.

If you would rather not post publicly, please contact us directly
using the support form.

We appreciate your feedback.

The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/home/agent/work/323e0f636337ffe3/_django_repo/tests/queries/tests.py", line 2817, in test_subquery_exclude_outerref
    self.assertTrue(qs.exists())
  File "/home/agent/work/323e0f636337ffe3/_django_repo/django/db/models/query.py", line 777, in exists
    return self.query.has_results(using=self.db)
  File "/home/agent/work/323e0f636337ffe3/_django_repo/django/db/models/sql/query.py", line 537, in has_results
    return compiler.has_results()
  File "/home/agent/work/323e0f636337ffe3/_django_repo/django/db/models/sql/compiler.py", line 1114, in has_results
    return bool(self.execute_sql(SINGLE))
  File "/home/agent/work/323e0f636337ffe3/_django_repo/django/db/models/sql/compiler.py", line 1144, in execute_sql
    cursor.execute(sql, params)
  File "/home/agent/work/323e0f636337ffe3/_django_repo/django/db/backends/utils.py", line 68, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/home/agent/work/323e0f636337ffe3/_django_repo/django/db/backends/utils.py", line 77, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/home/agent/work/323e0f636337ffe3/_django_repo/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
  File "/home/agent/work/323e0f636337ffe3/_django_repo/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/agent/work/323e0f636337ffe3/_django_repo/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.InternalError: internal error: column 6 not in input
```
The query is:
```sql
SELECT (1) AS "a" FROM "queries_jobresponsibilities"
WHERE EXISTS(
    SELECT V0."id", V0."description"
    FROM "queries_responsibility" V0
    WHERE NOT (V0."description" IN (
        SELECT U1."responsibility_id"
        FROM "queries_jobresponsibilities" U1
        WHERE U1."job_id" = "queries_jobresponsibilities"."job_id"
    ))
) LIMIT 1;
```
Schema:
```sql
CREATE TABLE "queries_job" ("id" integer NOT NULL PRIMARY KEY DEFAULT unique_rowid(), "name" varchar(20) NOT NULL UNIQUE);
CREATE TABLE "queries_jobresponsibilities" ("id" integer NOT NULL PRIMARY KEY DEFAULT unique_rowid(), "job_id" varchar(20) NOT NULL);
CREATE TABLE "queries_responsibility" ("id" integer NOT NULL PRIMARY KEY DEFAULT unique_rowid(), "description" varchar(20) NOT NULL UNIQUE);
ALTER TABLE "queries_jobresponsibilities" ADD COLUMN "responsibility_id" varchar(20) NOT NULL;
ALTER TABLE "queries_jobresponsibilities" ADD CONSTRAINT "queries_jobresponsibilities_job_id_e26ee158_fk_queries_job_name" FOREIGN KEY ("job_id") REFERENCES "queries_job" ("name");
ALTER TABLE "queries_jobresponsibilities" ADD CONSTRAINT "queries_jobresponsib_responsibility_id_154e9887_fk_queries_r" FOREIGN KEY ("responsibility_id") REFERENCES "queries_responsibility" ("description");
```

**Comment by timgraham** [comment]
The error message is similar to https://github.com/cockroachdb/cockroach/issues/39435. @RaduBerinde do you have an idea about this?

**Comment by RaduBerinde** [comment]
I think there is an outstanding open bug related to apply joins and subqueries, that might be what's happening here.

**Comment by RaduBerinde** [comment]
Ah, actually the one I'm thinking of would crash during execution of applyJoinNode, not during initial planning. Do you have the schema? I can try to reproduce with various table stats.

**Comment by rafiss** [comment]
I actually encountered a similar error when running the Hibernate test suite. See https://github.com/cockroachdb/cockroach/issues/50746

Also, I noticed it in these sentry reports
https://github.com/cockroachdb/cockroach/issues/48055
https://github.com/cockroachdb/cockroach/issues/50531

**Comment by timgraham** [comment]
If needed, I added the schema to this issue's description.

**Comment by RaduBerinde** [comment]
I have confirmed that this is the same with https://github.com/cockroachdb/cockroach/issues/50779, which was already fixed and backported to 19.2 and 20.1.

I was able to reproduce in v20.1.1 by adding statistics to make the U1 table larger (so that lookup anti-join is used):
```
ALTER TABLE queries_jobresponsibilities INJECT STATISTICS '[{"columns": ["id"], "created_at": "2018-01-01 1:00:00.00000+00:00", "row_count": 1000000, "distinct_count": 10000}]';
```
I checked the plan and indeed it looks like an invalid plan like in 50779. I confirmed that I can't repro on the current `release-20.1` branch.
